[FSDP] Move param/buffer name comp. to ctor for `ignored_modules`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #77000 [FSDP] Do not clone buffers; offload buffers to CPU if needed
* **#76994 [FSDP] Move param/buffer name comp. to ctor for `ignored_modules`**

Generating ignored parameter names in the full post state dict hook was not correct in general. This moves the name computation back to the constructor.

I think that the names must be computed in the constructor. Suppose that we have a `k`-level deep FSDP module hierarchy and that we are at level `i`.
- The parameters at level `i` have clean names starting with `p_1` + `"."` + ... + `"."` + `p_{i-1}`, where `p_j` is the wrapped submodule's name from level `j`.
- If the full post state dict hook, module `i` can only recurse further downward and not back up. In other words, it can only work with `p_i`, ..., `p_k` but not `p_1`, ..., `p_{i-1}`.
- This PR's approach is to have each FSDP module compute names but then use the root FSDP module's names for all FSDP modules in the hierarchy since only the root has the fully prefixed names.

Differential Revision: [D36271737](https://our.internmc.facebook.com/intern/diff/D36271737)

Commit:
[FSDP] Move param/buffer name comp. to ctor for `ignored_modules`

[ghstack-poisoned]

Commit:
Update on "[FSDP] Move param/buffer name comp. to ctor for `ignored_modules`"




[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -546,10 +546,24 @@ def __init__(
-        ignored_params = self._get_ignored_parameters()
+        # Validate the ignored modules and derive the ignored parameters and
+        # buffers
+        ignored_modules = self._get_ignored_modules(module, ignored_modules)
```
Maybe add a comment / warning in docstring in addition to this just to make sure it is completely clear to user that things might break if the parameter name gets swapped out underneath?

awgu replied:
Added a warning in the docstring.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -728,18 +742,20 @@ def _init_reshard_after_forward(self):
+        We include the child modules of modules in ``_ignored_modules`` to be
+        more intuitive since ignoring a module should ignore its child modules
+        as well, and we exclude :class:`FullyShardedDataParallel` instances
+        since ``self`` may be the intended root instance that manages them.
```
I feel like if a `module` contains a nested FSDP instance and a user wants to ignore `module`, something has gone wrong because portions of it will still be sharded? Or do you anticipate this being a use case?

awgu replied:
Good point. I definitely have less experience with how users may want to use the `ignore_modules` argument, so I trust your intuition here. Though, I think this is something we can revisit since the only change here would be to error instead of ignore.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -765,68 +781,58 @@ def _get_ignored_modules(
         )
 
-    def _get_ignored_named_tensors(
+    def _get_ignored_buffers(
```
Do these need to be separate functions or can they just work together to return buffers and params since there seems to only be one use case?

awgu replied:
This can be combined with below, but since the next diff overrides this anyway, I will leave it as is.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -765,68 +781,58 @@ def _get_ignored_modules(
+        ignored_buffer_names = set(
+            buffer_to_buffer_name[b] for b in ignored_buffers
         )
+        assert ignored_param_names.isdisjoint(ignored_buffer_names), \
```
Do we need this assert even though we end up taking the union afterwards? Agreed it is for correctness, but also wondering the performance tradeoff. If it is done only once in ctor as I think it is, it should be fine though.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1531,35 +1540,26 @@ def _full_post_state_dict_hook(
+            # Clone non-ignored parameters before exiting the
+            # `_summon_full_params()` context
+            if clean_tensor_name(key) not in self._ignored_tensor_names:
+                if not getattr(state_dict[key], "_has_been_cloned", False):
```
combine the two if-statements?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3583,10 +3583,29 @@ def _get_param_name_to_param(
+    def module_fn(module, prefix, buffer_to_buffer_name):
+        if not isinstance(module, FullyShardedDataParallel):
+            for buffer_name, buffer in module.named_buffers(recurse=False):
+                is_shared_buffer = buffer in buffer_to_buffer_name
```
can shared buffers have distinct names? If so, might be worth returning both of them, just to be extra sure we never clone buffers?

rohan-varma reviewed (approved):
Doesn't this PR add support for not buffer cloning as well? Can we have only one PR, such as the next diff, that implements all buffer related changes needed (as I think this PR should just be for moving param name computation back to ctor)? 

Also have a couple questions inline, the most important probably being how do we handle shared buffers (and parameters as well). Can we have unittests that ensure shared buffers and parameters are not cloned when not expected to be?

awgu commented:
To clarify, this PR is handling **ignored modules'** parameters and buffers, the names of which were computed in the full post state dict hook previously. This moves that computation to the constructor.

The next PR in the stack addresses buffer issues for **all** modules.

Commit:
Update on "[FSDP] Move param/buffer name comp. to ctor for `ignored_modules`"


Generating ignored parameter names in the full post state dict hook was not correct in general. This moves the name computation back to the constructor.

I think that the names must be computed in the constructor. Suppose that we have a `k`-level deep FSDP module hierarchy and that we are at level `i`.
- The parameters at level `i` have clean names starting with `p_1` + `"."` + ... + `"."` + `p_{i-1}`, where `p_j` is the wrapped submodule's name from level `j`.
- If the full post state dict hook, module `i` can only recurse further downward and not back up. In other words, it can only work with `p_i`, ..., `p_k` but not `p_1`, ..., `p_{i-1}`.
- This PR's approach is to have each FSDP module compute names but then use the root FSDP module's names for all FSDP modules in the hierarchy since only the root has the fully prefixed names.

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36271737).

awgu commented:
Failure is due to `test_variant_consistency_eager_eig_cuda_complex64`, which is unrelated.

awgu commented:
@pytorchbot merge this please
